capreolus/searcher/special.py

```diff
@@ -0,0 +1,196 @@
+        self.write_trec_run(preds=train_run, outfn=final_runfn, mode="wt")
+
+        # dev and test
+        dev_test_urls = [
+            "https://msmarco.blob.core.windows.net/msmarcoranking/top1000.dev.tar.gz",
+            "https://msmarco.blob.core.windows.net/msmarcoranking/top1000.eval.tar.gz",
+        ]
+        runs = {}
+        for url in dev_test_urls:
+            extract_file_name = self.get_fn_from_url(url)
+            extract_dir = self.benchmark.collection.download_and_extract(url, tmp_dir, expected_fns=extract_file_name)
+            runs.update(self.convert_to_trec_runs(extract_dir / extract_file_name, style="eval"))
+        self.write_trec_run(preds=runs, outfn=final_runfn, mode="a")
+
+        with open(final_donefn, "wt") as f:
+            print("done", file=f)
+        return output_path
+
+
+@Searcher.register
+class MsmarcoPsgBm25(BM25, MsmarcoPsgSearcherMixin):
+    module_name = "msmarcopsgbm25"
+    dependencies = [
+        Dependency(key="benchmark", module="benchmark", name="msmarcopsg"),
+        Dependency(key="index", module="index", name="anserini"),
+    ]
+    config_spec = BM25.config_spec + [
+        ConfigOption("tripleversion", "small", "version of triplet.qid file, small, large.v1 or large.v2"),
+    ]
+
+    def _query_from_file(self, topicsfn, output_path, config):
+        final_runfn = os.path.join(output_path, "searcher")
+        final_donefn = os.path.join(output_path, "done")
+        if os.path.exists(final_donefn):
+            return output_path
+
+        output_path.mkdir(exist_ok=True, parents=True)
+        tmp_dir = self.get_cache_path() / "tmp"
+        tmp_topicsfn = tmp_dir / os.path.basename(topicsfn)
+        tmp_output_dir = tmp_dir / "BM25_results"
+        tmp_output_dir.mkdir(exist_ok=True, parents=True)
+
+        train_runs = self.download_and_prepare_train_set(tmp_dir=tmp_dir)
+        if not os.path.exists(tmp_topicsfn):
+            with open(tmp_topicsfn, "wt") as fout:
+                with open(topicsfn) as f:
+                    for line in f:
+                        qid, title = line.strip().split("\t")
+                        if qid not in self.benchmark.folds["s1"]["train_qids"]:
+                            fout.write(line)
+
+        super()._query_from_file(topicsfn=tmp_topicsfn, output_path=tmp_output_dir, config=config)
+        dev_test_runfile = tmp_output_dir / "searcher"
+        assert os.path.exists(dev_test_runfile)
+
+        # write train and dev, test runs into final searcher file
+        Searcher.write_trec_run(train_runs, final_runfn)
+        with open(dev_test_runfile) as fin, open(final_runfn, "a") as fout:
+            for line in fin:
+                fout.write(line)
+
+        with open(final_donefn, "w") as f:
+            f.write("done")
+        return output_path
+
+
+# todo: make this another type of "Module" (e.g. DPR Module)
+@Searcher.register
+class StaticTctColBertDev(Searcher, MsmarcoPsgSearcherMixin):
```

Comment:
Can you add a short docstring here also? I guess this downloads an existing run file for TctColBert?